Enhancements for the JPA-related content (#31297)
quarkusio/quarkus: docs/src/main/asciidoc/security-jpa-concept.adoc

For an example of practical use of Basic authentication and JPA, see the xref:security-basic-authentication-tutorial.adoc[Secure a Quarkus application with Basic authentication and JPA] tutorial.

== JPA entity specification

Quarkus security offers a JPA integration to collect usernames, passwords, and roles, and store them into JPA database entities.

The following JPA entity specification demonstrates how users' information needs to be stored in a JPA entity and properly mapped so that Quarkus can retrieve this information from a database.

* The `@UserDefinition` annotation must be present on a JPA entity, no matter if link:https://quarkus.io/guides/hibernate-orm-panache[simplified Hibernate ORM with Panache] is used or not.

* The `@Username` and `@Password` field types are always `String`.

* The `@Roles` field must either be `String`, `Collection<String>`, or a `Collection<X>`, where `X` is an entity class with a single `String` field annotated as `@RolesValue`.

* Each `String` role element type is parsed as a comma-separated list of roles.

> 

Reviewer: Please copy the section from the Basic Auth and JPA tutorial over here

Reviewer: It will cause some duplication but but the Basic Auth and JPA tutorial is about getting started fast with Quarkus sec with the help of Basic Auth and JPA, while the purpose of this doc is to offer a complete JPA concept

Author: OK @sberyozkin . Adding now. Will rebase and squash.

Author: Done and done. See below from row 28.

@Entity
public class CustomPasswordUserEntity {
    @Id
    @GeneratedValue
    public Long id;

    @Column(name = "username")
    @Username
    public String name;

    @Column(name = "password")
    @Password(value = PasswordType.CUSTOM, provider = CustomPasswordProvider.class)
    public String pass;

    @Roles
    public String role;
}

public class CustomPasswordProvider implements PasswordProvider {
    @Override
    public Password getPassword(String pass) {
        byte[] digest = DatatypeConverter.parseHexBinary(pass);
        return SimpleDigestPassword.createRaw(SimpleDigestPassword.ALGORITHM_SIMPLE_DIGEST_SHA_256, digest);
    }
}
----

[TIP]
====
For quick creation of a hashed password, use `String BcryptUtil.bcryptHash(String password)`, which defaults to creating a random salt and hashing in ten iterations.
This method also allows specifying the desired amount of iterations and the salt used.
====

[WARNING]
====
Never store passwords as a plain text for applications already running in production.

However, it is possible to store passwords as a plain text with the `@Password(PasswordType.CLEAR)` annotation when operating in a test environment.

> ====

Reviewer: Please add Reeferences section here, every sec doc should link back to the main security overview concept and other related concept for the ease of navigation

Author: Added ;) Good idea

[id="security-jpa-concept"]
= Quarkus Security with JPA
include::_attributes.adoc[]

> :categories: security,jpa

Reviewer 2: Is this a good topic to help a developer get started with Quarkus? If so: 

```suggestion
:categories: security,jpa,getting-started
```

Author: @sberyozkin WDYT?

Reviewer: Hi @michelle-purcell @MichalMaler, on its own is a fairly technical concept, it is not really something that new users should be forwarded to when starting with Quarkus. It does make sense to show JPA in the getting started with quarkus security (Basic plus JPA) just because saving the users info in DB is better than directly in application.properties, but IMHO on its own it is not a getting started material :-). Perhaps at a later stage when a reactive JPA gets supported we can consider how to make it move visible, thanks

Reviewer 2: @sberyozkin - Thanks for explaining. Sounds good 👍

> :categories: security,jpa

Reviewer 2: Same comment here about `jpa` too.

> JPA is an identity provider, similar to xref:[JDBC], suitable for use with the xref:security-basic-authentication-concept.adoc[Basic] and xref:security-authentication-mechanisms-concept.adoc#form-auth[Form-based] Quarkus security mechanisms, which require a combination of username and password credentials.

Reviewer 2: Broken xref. See preview: 

![image](https://user-images.githubusercontent.com/92924207/220120098-3ee1da47-8b70-4922-8bcc-337c706ba972.png)

> JPA is an identity provider, similar to xref:[JDBC], suitable for use with the xref:security-basic-authentication-concept.adoc[Basic] and xref:security-authentication-mechanisms-concept.adoc#form-auth[Form-based] Quarkus security mechanisms, which require a combination of username and password credentials.

Reviewer 2: Fix xref.

> JPA is an identity provider, similar to xref:security-jdbc.adoc[JDBC], suitable for use with the xref:security-basic-authentication-concept.adoc[Basic] and xref:security-authentication-mechanisms-concept.adoc#form-auth[Form-based] Quarkus security mechanisms, which require a combination of username and password credentials.

Reviewer 2: I think the _**s**_ in security should be uppercase here and elsewhere when referring to the noun "Quarkus Security", which is the name of the Quarkus framework for security.

Author: Great catch @michelle-purcell . Fixing here and on some other places too.

The JPA `IdentityProvider` creates a `SecurityIdentity` instance, which is used during user authentication to verify and authorize access requests making your Quarkus application secure.

For an example of practical use of Basic authentication and JPA, see the xref:security-basic-authentication-tutorial.adoc[Secure a Quarkus application with Basic authentication and JPA] tutorial.

== JPA entity specification

Quarkus security offers a JPA integration to collect usernames, passwords, and roles, and store them into JPA database entities.

The following JPA entity specification demonstrates how users' information needs to be stored in a JPA entity and properly mapped so that Quarkus can retrieve this information from a database.

> * The `@UserDefinition` annotation must be present on a JPA entity, regardless of whether link:https://quarkus.io/guides/hibernate-orm-panache[simplified Hibernate ORM with Panache] is used or not.

Reviewer 2: @MichalMaler - Sorry just noticed this now while verifying single-sourced content on downstream docs. Should this be an xref? Thinking if not an xref, then later it might present a version mismatch as the link always points to the latest.

import io.quarkus.security.jpa.Roles;
import io.quarkus.security.jpa.UserDefinition;
import io.quarkus.security.jpa.Username;

@Entity
@Table(name = "test_user")
@UserDefinition <1>
public class User extends PanacheEntity {
    @Username <2>
    public String username;
    @Password <3>
    public String password;
    @Roles <4>
    public String role;

    /**
     * Adds a new user to the database
     * @param username the username
     * @param password the unencrypted password (it will be encrypted with bcrypt)
     * @param role the comma-separated roles
     */
    public static void add(String username, String password, String role) { <5>
        User user = new User();
        user.username = username;
        user.password = BcryptUtil.bcryptHash(password);
        user.role = role;
        user.persist();
    }
}

----

The `security-jpa` extension initializes only if a single entity is annotated with `@UserDefinition`.

<1> The `@UserDefinition` annotation must be present on a single entity, either a regular Hibernate ORM entity or a Hibernate ORM with a Panache entity.
<2> Indicates the field used for the username.
<3> Indicates the field used for the password.

> By default, it uses bcrypt-hashed passwords, and you can configure it to use plain text or custom passwords.

Reviewer 2: Clarify what the 'it' is:
```suggestion
By default, `security-jpa` uses bcrypt-hashed passwords, and you can also configure plain text or custom passwords instead.
```

<4> This indicates the comma-separated list of roles added to the target principal representation attributes.

> <5> This method allows us to add users while hashing passwords with the proper `bcrypt` hash.

Reviewer 2: Suggestion:Change _**us**_ to **_you_** to be consistent with previous statements.

== JPA entity as storage of roles

Use the following example to store roles inside another JPA entity:

[source,java]
----
@UserDefinition
@Table(name = "test_user")
@Entity
public class User extends PanacheEntity {
    @Username
    public String name;

    @Password
    public String pass;

    @ManyToMany
    @Roles
    public List<Role> roles = new ArrayList<>();
}

@Entity
public class Role extends PanacheEntity {

    @ManyToMany(mappedBy = "roles")
    public List<ExternalRolesUserEntity> users;

    @RolesValue
    public String role;
}
----

== Password storage and hashing

> Regarding password storage and hashing management from a developer perspective, you can decide if you want to keep the settings as it is and use the Quarkus default or hash passwords manually.

Reviewer 2: Suggestion to simplify if not breaking the technical meaning:

```suggestion
When developing applications with Quarkus, you can decide how to manage password storage and hashing. You can choose to keep the default password and hashing settings of Quarkus, or you can hash passwords manually.
```

With the default option, passwords are stored and hashed with https://en.wikipedia.org/wiki/Bcrypt[bcrypt] under the
https://en.wikipedia.org/wiki/Crypt_(C)[Modular Crypt Format] (MCF).
While using MCF, the hashing algorithm, iteration count, and salt are stored as a part of the hashed value.
As such, we do not need dedicated columns to keep them.

[NOTE]
====
In cryptography, a salt is a name for random data used as an additional input to a one-way function that hashes data, a password, or a passphrase.
====

For manual password hashing, create a class that implements the `CustomPasswordProvider`as shown in the example below.

> This is an example of providing a custom password provider using the SHA256 hashing algorithm.

Reviewer 2: ```suggestion
The following snippet shows how to set a custom password provider that uses the SHA256 hashing algorithm.
```

[source,java]
----
@UserDefinition
@Table(name = "test_user")
@Entity
public class CustomPasswordUserEntity {
    @Id
    @GeneratedValue
    public Long id;

    @Column(name = "username")
    @Username
    public String name;

    @Column(name = "password")
    @Password(value = PasswordType.CUSTOM, provider = CustomPasswordProvider.class)
    public String pass;

    @Roles
    public String role;
}

public class CustomPasswordProvider implements PasswordProvider {
    @Override
    public Password getPassword(String pass) {
        byte[] digest = DatatypeConverter.parseHexBinary(pass);
        return SimpleDigestPassword.createRaw(SimpleDigestPassword.ALGORITHM_SIMPLE_DIGEST_SHA_256, digest);
    }
}
----

[TIP]
====
For quick creation of a hashed password, use `String BcryptUtil.bcryptHash(String password)`, which defaults to creating a random salt and hashing in ten iterations.
This method also allows specifying the desired amount of iterations and the salt used.
====

[WARNING]
====

> Never store passwords as a plain text for applications already running in production.

Reviewer 2: Remove article **_a_**. Suggestion: 

```suggestion
For applications running in a production environment, do not store passwords as plain text.
```

> However, it is possible to store passwords as a plain text with the `@Password(PasswordType.CLEAR)` annotation when operating in a test environment.

Reviewer 2: Remove unrequired article _a_.

```suggestion
However, it is possible to store passwords as plain text with the `@Password(PasswordType.CLEAR)` annotation when operating in a test environment.
```

====

> == References

Reviewer 2: Links LGTM on preview 💯
